Serialize content of linked style sheets with URL replacement

szewai opened the pull request:
#### c72276aee6201a3ed519b5031f195acab04479a7
<pre>
Serialize content of linked style sheets with URL replacement
<a href="https://bugs.webkit.org/show_bug.cgi?id=264308">https://bugs.webkit.org/show_bug.cgi?id=264308</a>
<a href="https://rdar.apple.com/118029249">rdar://118029249</a>

Reviewed by Ryosuke Niwa.

We need to replace URLs in external style sheets when saving them to disk.

API test: WebArchive.SaveResourcesLink

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::cssTextWithReplacementURLs):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::styleSheetContentWithReplacementURLs const):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::textContentWithReplacementURLs const):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/270329@main">https://commits.webkit.org/270329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d64734559056f2ba4a458bc3202e1d146e5b6b4f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25156 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3697 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27271 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23081 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1133 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23326 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25399 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2721 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21709 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27850 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2407 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22649 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28774 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22996 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26597 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/652 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3707 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6038 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2798 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2693 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -616,6 +617,32 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, Lo
+                HashMap<String, String> uniqueSubresourcesInElement;
+                for (auto [urlString, path] : uniqueSubresources) {
+                    if (subresourceURLs.contains(URL { urlString })) {
+                        // The linked flle is placed in subresource directory as other subresource files.
```
Nit: flle -> file
Nit: as -> with.

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -616,6 +617,32 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, Lo
+            if (!subresourcesDirectoryName.isNull() && is<HTMLLinkElement>(node)) {
+                auto& element = downcast<HTMLLinkElement>(node.get());
+                auto index = subresources.findIf([&](auto& resource) {
+                    return resource->url() == element.href();
```
Not every link element is a stylesheet.
We should probably only do this when sheet() returns a non-null value.
e.g. rel=preload resources should probably be ignored.

szewai replied:
In this patch we only replace file content when `m_sheet` is non-null, as `HTMLLinkElement::sytleSheetContentWithReplacementURLs` returns null when `m_sheet` is null and we don't overwrite the resource when content string is null.

rniwa replied:
What I meant is that we should probably do that check here instead of always calling `styleSheetContentWithReplacementURLs`. By the way, there is a typo in that function name: `sytle`.

szewai replied:
Oh will correct; and I will add a check here

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -616,6 +617,32 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, Lo
+                    }
+                }
+
+                auto contentString = element.sytleSheetContentWithReplacementURLs(uniqueSubresourcesInElement);
```
Nit: sytle -> style

szewai committed:
Serialize content of linked style sheets with URL replacement
https://bugs.webkit.org/show_bug.cgi?id=264308
rdar://118029249

Reviewed by Ryosuke Niwa.

We need to replace URLs in external style sheets when saving them to disk.

API test: WebArchive.SaveResourcesLink

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::cssTextWithReplacementURLs):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::styleSheetContentWithReplacementURLs const):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::textContentWithReplacementURLs const):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/270329@main
